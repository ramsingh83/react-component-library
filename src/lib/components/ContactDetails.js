import React from "react";
import "./ContactDetails.css";

const ContactDetails = ({
  label,
  placeholder,
  email,
  handleOnInputChanged,
  invalid,
  required
}) => (
  <div className="form-item">
    <label htmlFor="email-id">
      <span className={`input-label ${required ? 'element-required' : ''}`}>{label}</span>
      <input
        id="email-id"
        type="text"
        className={`form-input ${invalid ? 'invalid' : ''}`}
        aria-describedby="email-error"
        aria-label={label}
        placeholder={placeholder}
        onChange={handleOnInputChanged}
        value={email}
        maxLength="128"
        aria-required="true"
        aria-invalid={!!invalid} />
    </label>
    <div id="email-error" className="error-ifo"></div>
  </div>
);

export default ContactDetails;
