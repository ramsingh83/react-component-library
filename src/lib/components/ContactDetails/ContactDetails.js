import React, { useState, useLayoutEffect } from 'react';
import PropTypes from 'prop-types';

const ContactDetails = ({
  title,
  email,
  phone,
  fields,
  validateInput,
  setValues
}) => {
  const [contacts, setContacts] = useState({
    email,
    phone,
    emailError: '',
    phoneError: ''
  });

  useLayoutEffect(() => {
    setValues(contacts);
  }, [contacts]);

  const [firstRender, setFirstRender] = useState(true);

  const validate = () => {
    const values = { ...contacts };
    fields.forEach((field) => {
      const pattern = new RegExp(field.pattern);
      const value = values[field.key];
      if (!value && field.required) {
        values[`${field.key}Error`] = field.emptyError;
      } else if (value && pattern && !pattern.test(value)) {
        values[`${field.key}Error`] = field.invalidError;
      } else {
        contacts[`${field.key}Error`] = '';
      }
    });
    setContacts(values);
  };

  useLayoutEffect(() => {
    setFirstRender(false);
    if (!firstRender) {
      validate();
    }
  }, [validateInput]);

  const setError = key => contacts[`${key}Error`];

  const handleOnInputChanged = (event) => {
    event.preventDefault();
    const { name, value } = event.target;
    const inputDetails = fields.find(input => input.key === name);
    if (!value) {
      if (inputDetails.required) {
        contacts[`${name}Error`] = inputDetails.emptyError;
      } else {
        contacts[`${name}Error`] = '';
      }
    } else {
      contacts[`${name}Error`] = '';
    }

    setContacts({ ...contacts, [name]: value });
  };

  const handleFocusOut = (event) => {
    event.preventDefault();
    const { name, value } = event.target;
    const inputs = fields.find(input => input.key === name);
    const pattern = new RegExp(inputs.pattern);
    if (value) {
      if (pattern && !pattern.test(value)) {
        contacts[`${name}Error`] = inputs.invalidError;
      } else {
        contacts[`${name}Error`] = '';
      }
    } else if (inputs.required) {
      contacts[`${name}Error`] = inputs.emptyError;
    } else {
      contacts[`${name}Error`] = '';
    }
    setContacts({ ...contacts, [name]: value });
  };

  return (
    <fieldset className="contact-details">
      <legend>{title}</legend>
      <div>
        {
          fields.map(field => (
            <React.Fragment key={field.key}>
              <div className="form-item">
                <label htmlFor={field.key} className="input-label">
                  {field.label}
                  {field.required ? <span className="mandatory">*</span> : ''}
                  <input
                    name={field.key}
                    type="text"
                    className={`form-input ${setError(field.key) ? 'invalid' : ''}`}
                    aria-describedby={`${field.key}-error`}
                    aria-label={field.label}
                    placeholder={field.placeholder}
                    autoComplete={field.autoComplete || 'off'}
                    onChange={handleOnInputChanged}
                    onBlur={handleFocusOut}
                    value={contacts[field.key] || ''}
                    maxLength={field.maxLength}
                    minLength={field.minLength}
                    aria-required="true"
                    aria-invalid={!!setError(field.key)} />
                </label>
                <div id={`${field.key}-error`} className="error-info">{setError(field.key)}</div>
              </div>
            </React.Fragment>
          ))
        }
      </div>
    </fieldset>
  );
};

ContactDetails.propTypes = {
  title: PropTypes.string,
  email: PropTypes.string,
  phone: PropTypes.string,
  fields: PropTypes.arrayOf(PropTypes.shape({})),
  validateInput: PropTypes.bool,
  setValues: PropTypes.func
};

export default ContactDetails;
