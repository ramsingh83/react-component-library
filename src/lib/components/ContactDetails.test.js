import React from "react";
import ContactDetails from "./ContactDetails";
import renderer from "react-test-renderer";

describe("ContactDetails", () => {
  it("renders properly", () => {
    const tree = renderer
      .create(<ContactDetails label="Email" placeholder="name@example.com" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
