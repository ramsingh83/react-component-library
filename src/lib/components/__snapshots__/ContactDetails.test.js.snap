// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactDetails renders properly 1`] = `
<div
  className="form-item"
>
  <label
    htmlFor="email-id"
  >
    <span
      className="input-label "
    >
      Email
    </span>
    <input
      aria-describedby="email-error"
      aria-invalid={false}
      aria-label="Email"
      aria-required="true"
      className="form-input "
      id="email-id"
      maxLength="128"
      onChange={undefined}
      placeholder="name@example.com"
      type="text"
      value={undefined}
    />
  </label>
  <div
    className="error-ifo"
    id="email-error"
  />
</div>
`;
